Delete pool before getting Node Address unsuccessfully

FengyunPan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:

**Which issue this PR fixes** _(optional, in `fixes #<issue number>(, #<issue_number>, ...)` format, will close that issue when PR gets merged)_: fixes #

**Special notes for your reviewer**:

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

``` release-note
```

When calls getAddressByName(), k8s returns error without deleting
the pool of openstack. So many lb-pool will be left over. We should
delete it.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32120)

<!-- Reviewable:end -->

mikedanese commented:
cc @kubernetes/sig-openstack

feiskyer commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -861,7 +864,9 @@ func (lb *LbaasV1) EnsureLoadBalancer(clusterName string, apiService *api.Servic
 		}).Extract()
 		if err != nil {
-			pools.Delete(lb.network, pool.ID)
+                        if pool != nil {
```
This check is unnecessary, since above codes already refer `pool.ID`.

feiskyer commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -852,6 +852,9 @@ func (lb *LbaasV1) EnsureLoadBalancer(clusterName string, apiService *api.Servic
 	for _, host := range hosts {
 		addr, err := getAddressByName(lb.compute, host)
 		if err != nil {
+                        if pool != nil {
```
And I think this check is also unnecessary, `pools.Create` should return a none-nil `pool` on success call (Correct me if I'm wrong).

FengyunPan replied:
If hosts have multiple names, pool.delete will be called repeatedly. But pool just only one.
If pool.delete has be called, we should add check.

feiskyer replied:
> pool.delete will be called repeatedly

why this? if error occurred, it will jump out by `return nil, err` below.

FengyunPan replied:
Oh, I see, thank you very much. I will delete the check.

feiskyer commented:
LGTM.

anguslees commented:
LGTM.

anguslees commented:
This is probably a candidate for 1.3 release, if anyone is watching who can manipulate labels.

anguslees commented:
(ping @mikedanese)

mikedanese commented:
@k8s-bot ok to test

mikedanese commented:
Needs a gofmt or a govet or something. This is not passing verification.

Commit:
Delete pool before getting Node Address unsuccessfully

When calls getAddressByName(), k8s returns error without deleting
the pool of openstack. So many lb-pool will be left over. We should
delete it.

FengyunPan commented:
@mikedanese Thank you very much, I have created a new patch for gofmt.

anguslees commented:
Just so the two PRs know about each other: https://github.com/kubernetes/kubernetes/pull/33276 effectively obsoletes this change by making k8s recover from this half-created situation more cleverly.

(#32120 is extremely minimal fix, and my original intention was to get it into 1.4 before release.  I think it's still worth adding this PR to a 1.4.x point release if that's something we do.)

FengyunPan commented:
(ping @mikedanese)

suonto commented:
It is potentially dangerous to delete the pool upon an error in creation of any of the members. Let me explain this by an example: We have a 1000 node cluster. So for a service, we create a load balancer, along with a listener + one pool and 1000 members, one for each node. Now, primary master fails, master failover occurs and EnsureLoadBalancer is called. If nodeNames -list has not been updated, getting address for ex-primary master will get ErrNotFound. Also, if any other node is down, we get ErrNotFound. If we delete the pool in that occasion, it means that the other 999 nodes can't serve our service anymore either. Instead, we should do something else, like this:
```
addr, err := getAddressByName(lbaas.compute, types.NodeName(nodeName))
if err != nil {
	if err == ErrNotFound {
		// Node failure, do not create member
		glog.Warningf("Failed to create LB pool member for node %s: %v", nodeName, err)
		continue
	} else {
		// Possible cleanup
		return nil, fmt.Errorf("Error getting address for node %s: %v", nodeName, err)
	}
}
```
A good candidate for cleanup strategy is to defer delete calls to any objects created by EnsureLoadBalancer (as @anguslees suggested in #33276 discussion) and disarm them on success.
Also, we might consider member creation to also fail critically if all address queries get ErrNotFound (and perform cleanup).

FengyunPan commented:
@suonto Thank you for your comments, the member and pool are the same, I will fix it.

anguslees commented:
#33276 has merged, so this change is unnecessary for master/v1.5.  It is reasonable to merge this minimal PR into v1.4, if we do bugfix point releases.
